src/uct/ib/rc/base/rc_iface.c

```diff
@@ -502,7 +503,11 @@ ucs_status_t uct_rc_iface_init_rx(uct_rc_iface_t *iface,
     srq_init_attr.srq_context    = iface;
     srq                          = ibv_create_srq(pd, &srq_init_attr);
     if (srq == NULL) {
-        ucs_error("ibv_create_srq() failed: %m");
+        ucs_string_buffer_appendf(&msg, "ibv_create_srq() failed: %m");
+        if (errno == ENOMEM) {
+            ucs_sys_check_memlock_limit_append_msg(&msg);
+        }
+        ucs_error("%s", ucs_string_buffer_cstr(&msg));
```

Comment:
seems the pattern above repeats in : uct_rc_iface_init_rx, uct_ib_mlx5_create_cq, uct_ib_mlx5dv_check_dc, uct_ib_verbs_create_cq
Can we unite the common code?

Reply:
I can imagine two ways of unification:
1. Just introduce the common function only for this case. I mean something like `void ucs_sys_report_error_with_memlock_limit_check(const char *fmt, ...)` that will look just like this:
    ```c
    void ucs_sys_report_error_with_memlock_limit_check(const char *fmt, ...) {
        UCS_STRING_BUFFER_ONSTACK(msg, 256);
        va_list ap;
        va_start(ap, fmt);
        ucs_string_buffer_appendf(&msg, fmt, ap);
        va_end(ap);
        if (errno == ENOMEM) {
            ucs_sys_check_memlock_limit_append_msg(&msg);
        }
        ucs_error("%s", ucs_string_buffer_cstr(&msg));
    }
    ```
    But IMO this is not an improvement. This function just encapsulates the one case into a separate function but spreads code across the different files. 
2. Include the common way of analyzing the `errno` value to the ucs_error/ucs_warning/... . This is a more common change, but it will unify all `errno` handling in one place. I think it is a good idea, but I don't know do we have cases that may be broken by this design.

WDYT?

Reply:
```
void ucs_sys_report_error_with_memlock_limit_check(const char *func_name, int errno_code) {
}
```

Reply:
I discovered that in some places that we want to unify the UCX reports not only strings like `ibv_create_srq() failed:` but also formatted string, e.g. `ibv_create_cq(cqe=%d) failed:`.

So to not reduce the amount of logging information, I think we need to choose something like the first option proposed by me above. WDYT?

Reply:
maybe something like:
```c

static void uct_ib_mem_lock_limit_msg(const char *message, int sys_errno)
{
       size_t memlock_limit;
       ucs_status_t status;

       if (sys_errno == ENOMEM) {
            status = ucs_sys_get_memlock_rlimit(&memlock_limit);
            if (status == UCS_OK) {
                ucs_error("%s failed: %s. Please set max locked memory ...", 
                          strerror(sys_errno), memlock_limit/UCS_KBYTE);
                return;
            }
       }

       ucs_error("%s failed: %s", message, strerror(sys_errno));
}
...

uct_ib_mem_lock_limit_msg("ibv_create_cq(cqe=1)", errno)

...

char message[128];
ucs_snprintf_safe(message, sizeof(message), "mlx5dv_create_cq(cqe=%d)", cqe);
uct_ib_mem_lock_limit_msg("message", errno);

...

uct_ib_mem_lock_limit_msg("ibv_create_srq()", errno);

```